### PR TITLE
Ruler: Add support for arbitrary extra URL parameters to Alertmanager client's OAuth config

### PR DESCRIPTION
**Pull request by alexweav**
#### What this PR does

This PR allows you to specify arbitrary URL parameters to send alongside OAuth token requests, when the Alertmanager client is configured in OAuth mode.

Some OAuth providers require extra fields here that are not normally part of the spec - for example, `audience` is a common one. The Go stdlib followed a similar convention for such fields: https://github.com/golang/oauth2/issues/216 where arbitrary extra URL values are allowed.

#### Which issue(s) this PR fixes or relates to

n/a

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/util/validation/limits_map.go [hunk 1]
```diff
@@ -9,15 +9,19 @@ import (
 	"gopkg.in/yaml.v3"
 )
 
-// LimitsMap is a generic map that can hold either float64 or int as values.
-type LimitsMap[T float64 | int] struct {
+// LimitsMap is a flag.Value implementation that looks like a generic map, holding float64s, ints, or strings as values.
+type LimitsMap[T float64 | int | string] struct {
```
Comment (alexweav) [inline pkg/util/validation/limits_map.go:13]: So far, the only place where we really use map flags is in Limits, that's why this was called LimitsMap. But, nothing in this file is really specific to limits.

This new use case isn't really a limit, but I felt it gives a better configuration experience to be consistent with existing flags here, rather than invent a second map format that behaves differently.

I've kept the name `LimitsMap` and package as i feel it's _close enough_ to being descriptive and would muddy the diff to include renames/moves. I also think there's a valid case for moving this to `dskit`'s `flagext` package in the future anyway, so it didn't make sense to do a big rename yet.

I'm happy to change the name, or explore moving this to dskit, if reviewers would prefer. Or, this can be left to a subsequent PR.

File: pkg/util/validation/limits_map.go [hunk 2]
```diff
@@ -73,6 +81,7 @@ func (m LimitsMap[T]) updateMap(newMap map[string]T) error {
 		}
 	}
 
+	clear(m.data)
```
Comment (alexweav) [inline pkg/util/validation/limits_map.go:84]: drive-by bugfix i happened to notice. added corresponding test.

**Review by tacole02 — APPROVED** [review]
👍

**Review by francoposa — APPROVED** [review]
LGTM - I think putting it here is probably the right route for now.

I will say that even though it's a minor bit of code I don't think we need to be in the habit of validating something like these endpoint param as `validateOAuth2EndpointParam` does. Empty query param values are valid for HTTP, and since this config is for for connecting to/through some external service, it's not really our concern to tell the operator what is or is not valid.
